### PR TITLE
Add an example of calculating the SHA-256 digest of a file using ring

### PR DESCRIPTION
**Pull request by luser**
Hopefully this looks sensible! I put it in basics because I didn't know where else to put it.

### CODE REVIEW

File: src/basics.md [hunk 1]
```diff
@@ -773,6 +831,8 @@ fn run() -> Result<()> {
 [bitflags]: https://docs.rs/bitflags/
 [byteorder-badge]: https://badge-cache.kominick.com/crates/v/byteorder.svg?label=byteorder
 [byteorder]: https://docs.rs/byteorder/
+[data-encoding-badge]: https://badge-cache.kominick.com/crates/v/data-encoding.svg?label=data-encoding
+[data-encoding]: https://github.com/ia0/data-encoding
```
Comment (budziq) [inline src/basics.md]: please copy the same link as the one in encoding.md

File: src/encoding.md [hunk 2]
```diff
@@ -553,7 +553,7 @@ fn run() -> Result<()> {
 [base64-badge]: https://badge-cache.kominick.com/crates/v/base64.svg?label=base64
 [base64]: https://docs.rs/base64/
 [data-encoding-badge]: https://badge-cache.kominick.com/crates/v/data-encoding.svg?label=data-encoding
-[data-encoding]: https://github.com/ia0/data-encoding
+[data-encoding]: https://docs.rs/data-encoding/
```
Comment (budziq) [inline src/encoding.md]: 👍

File: src/basics.md [hunk 3]
```diff
@@ -748,6 +749,63 @@ fn run() -> Result<()> {
...
+
+[![ring-badge]][ring] [![data-encoding-badge]][data-encoding] [![cat-filesystem-badge]][cat-filesystem]
+
+Writes some data to a file, then calculates the SHA-256 digest of
+the file's contents using [`digest::Context`].
+
+```rust
+# #[macro_use]
+# extern crate error_chain;
+
```
Comment (budziq) [inline src/basics.md]: please add one more `#` on the begining of line here so we don't have weird empty lines in final output

File: src/basics.md [hunk 4]
```diff
@@ -748,6 +749,63 @@ fn run() -> Result<()> {
...
+extern crate data_encoding;
+extern crate ring;
+
+use data_encoding::HEXUPPER;
+use ring::digest::{Context, SHA256};
+use std::fs::File;
+use std::io::{BufReader, Read, Write};
+
```
Comment (budziq) [inline src/basics.md]: please add one more `#` on the begining of line here so we don't have weird empty lines in final output

File: src/basics.md [hunk 5]
```diff
@@ -748,6 +749,63 @@ fn run() -> Result<()> {
...
+    let path = "file.txt";
+
+    let mut output = File::create(path)?;
+    write!(output, "We will generate a digest of this text")?;
+
+    let input = File::open(path)?;
+    let mut reader = BufReader::new(input);
+    let mut context = Context::new(&SHA256);
+
+    loop {
```
Comment (budziq) [inline src/basics.md]: Normally I would go with extracting the reading looping and digest into a yet another function taking io::Read and returning Result<Digest>.
But Its perfect as it is so I'll leave it up to you.

File: src/basics.md [hunk 6]
```diff
@@ -748,6 +749,63 @@ fn run() -> Result<()> {
 
 ```
 
+[ex-sha-digest]: #ex-sha-digest
+<a name="ex-sha-digest"></a>
+## Calculate the SHA-256 digest of a file
+
+[![ring-badge]][ring] [![data-encoding-badge]][data-encoding] [![cat-filesystem-badge]][cat-filesystem]
+
+Writes some data to a file, then calculates the SHA-256 digest of
```
Comment (budziq) [inline src/basics.md]: I would probably changed "digest" to linked  [`Digest`](https://docs.rs/ring/*/ring/digest/struct.Digest.html).

File: src/basics.md (hunk 5 above)
Comment (luser) [inline src/basics.md]: That's a great suggestion, I factored out a `sha256_digest_from_read` function which is a bit more copy-paste friendly, which I assume is part of the point of the cookbook. :)

Comment (budziq) [inline src/basics.md]: 👍

**Comment by budziq** [comment]
@luser nicely done. Merged

**Pull request closed without merging**